src/convolution/convolution_2d.jl

```diff
@@ -0,0 +1,1153 @@
+"""
+
+# ╔═╡ 98919284-8099-43c8-b123-d1476d27a88d
+img = rand(Gray, 5, 5)
+
+# ╔═╡ eb983abe-5899-431b-84a1-7d06715d75c4
+md"""
+Now that you know how images are built. Let's get to the fun part. Below is a dropdown with popular test images. Choose the one you like and let's start learning about filters.
+"""
+
+# ╔═╡ 4edbfca5-cb25-4093-9f4d-5cfa2aa8d77f
+@bind test_img_selected Select([
+	testimage("coffee") => "Coffee", 
+	testimage("house") => "House",
+	testimage("pirate") => "Pirate",
+	testimage("peppers") => "Peppers",
+	testimage("mandrill") => "Mandrill"
+	])
+
+# ╔═╡ ef0bea69-6dfd-4526-a96e-6607771660fe
+test_img = Gray.(test_img_selected)
+
+# ╔═╡ ebb63415-2738-47f3-b0ef-0fd6e6bac259
+md"""
+## Using Filter 
+
+Now to make things more interesting, we will add some code that applies a filter to this image.
+
+This code basically goes over all the pixels in the image and applies the mathematical operation of convolution, that is the values of the pixel and its surrounding pixels are multipled by the value of the filter matrix. 
+
+However, it's alright if you don't understand exactly how the code works."""
+
+# ╔═╡ 44d594ff-a25b-4455-be56-863345c67b68
+function convolve(img, filter)
+    img_row, img_col = size(img)
+    filter_row, filter_col = size(filter)
+
+    filtered = zeros(img_row-filter_row+1, img_col-filter_col+1)
+    filtered_row, filtered_col = size(filtered)
+
+    for i in 1:filtered_row
+        for j in 1:filtered_col
+            filtered[i,j] = sum(img[i:i+filter_row-1,j:j+filter_col-1].*filter)
+        end
+    end
+  
+    return filtered
+end
+
+# ╔═╡ d752805d-ba2c-4133-a556-46a1da734557
+md"""Now let's start with a simple filter. This is called a "Box Blur" and it has the simple task of blurring the image.
+
+Can you see it ?"""
+
+# ╔═╡ d729f8b4-05e2-4863-b8c4-39b37646c36b
+test_filter = [0.111 0.111 0.111; 0.111 0.111 0.111; 0.111 0.111 0.111]
+
+# ╔═╡ 8ae0248f-6eba-4941-85f4-b21be3c7e725
+let
+	height, width = size(test_img)
+	filtered_img = Gray.(convolve(test_img, test_filter))
+	padded_img = test_img[2:height-1, 2:width-1]
+	[padded_img filtered_img]
+end
+
+# ╔═╡ 19b49665-0382-4eb1-9c70-8295e0aa819b
+md"""You can also try other filters! 
+
+Choose a filter from the selection below and see what the filter matrix looks like and how the image changes. 
+
+Can you identify the difference between the right and left shift matrices ?"""
+
+# ╔═╡ adc154cf-7059-4f1d-9bac-56b9a93cc47f
+@bind test_filter_selected Select([
+	[0 0 0; 0 1 0; 0 0 0] => "Original Image",
+	[1 2 1; 2 4 2; 1 2 1] * 0.0625 => "Gaussian Blur", 
+	[0.111 0.111 0.111; 0.111 0.111 0.111; 0.111 0.111 0.111] => "Box Blur",
+	[-0.111 -0.111 -0.111; -0.111 0.89 -0.111; -0.111 -0.111 -0.111] => "Details",
+	[-1 -1 -1; -1 8 -1; -1 -1 -1] => "Ridge Detection",
+	[0 -1 0; -1 4 -1; 0 -1 0] => "Edge Detection",
+	[1 2 1; 0 0 0; -1 -2 -1] => "Vertical Edges",
```

Comment:
vertical & horizontal should be swapped (that is, the current vertical one, is actually detecting horizontal edges - I dont mean to swap the position in the list ;))